code/game/machinery/vending.dm

```diff
@@ -336,96 +299,66 @@
-			return
-
-		dat += "<h3>Select an item</h3>"
-		dat += "<div class='statusDisplay'>"
-		if(product_records.len == 0)
-			dat += "<font color = 'red'>No product loaded!</font>"
+/obj/machinery/vending/proc/build_inventory(list/productlist, hidden=0, req_coin=0, start_empty = null)
+	for(var/typepath in productlist)
+		var/amount = productlist[typepath]
+		if(isnull(amount))
+			amount = 0
+		var/atom/temp = new typepath(null)
+		var/datum/data/vending_product/R = new /datum/data/vending_product()
+		R.product_name = initial(temp.name)
+		R.product_path = typepath
+		if(!start_empty)
+			R.amount = amount
+		R.max_amount = amount
+		R.display_color = pick("firebrick", "dodgerblue", "royalblue", "indianred", "mediumaquamarine", "mediumpurple", "forestgreen", "goldenrod", "chocolate", "olivedrab")
+		if(hidden)
+			hidden_records += R
+		else if(req_coin)
+			coin_records += R
 		else
-			var/list/display_records = product_records
-			if(extended_inventory)
-				display_records = product_records + hidden_records
-			if(coin)
-				display_records = product_records + coin_records
-			if(coin && extended_inventory)
-				display_records = product_records + hidden_records + coin_records
-			dat += "<ul>"
-			for (var/datum/data/vending_product/R in display_records)
-				dat += "<li>"
-				if(R.amount > 0)
-					dat += "<a href='byond://?src=\ref[src];vend=\ref[R]'>Vend</a> "
-				else
-					dat += "<span class='linkOff'>Sold out</span> "
-				dat += "<font color = '[R.display_color]'><b>[sanitize(R.product_name)]</b>:</font>"
-				dat += " <b>[R.amount]</b>"
-				dat += "</li>"
-			dat += "</ul>"
-		dat += "</div>"
-		if(premium.len > 0)
-			dat += "<b>Coin slot:</b> "
-			if (coin)
-				dat += "[coin]&nbsp;&nbsp;<a href='byond://?src=\ref[src];remove_coin=1'>Remove</a>"
-			else
-				dat += "<i>No coin</i>&nbsp;&nbsp;<span class='linkOff'>Remove</span>"
-		if(istype(src, /obj/machinery/vending/snack))
-			dat += "<h3>Chef's Food Selection</h3>"
-			dat += "<div class='statusDisplay'>"
-			for (var/O in dish_quants)
-				if(dish_quants[O] > 0)
-					var/N = dish_quants[O]
-					dat += "<a href='byond://?src=\ref[src];dispense=[sanitize(O)]'>Dispense</A> "
-					dat += "<B>[capitalize(O)]: [N]</B><br>"
-			dat += "</div>"
-	user.set_machine(src)
-	if(seconds_electrified && !(stat & NOPOWER))
-		if(shock(user, 100))
-			return
-
-	var/datum/browser/popup = new(user, "vending", (name))
-	popup.set_content(dat)
-	popup.set_title_image(user.browse_rsc_icon(src.icon, src.icon_state))
-	popup.open()
-
+			product_records += R
 
-/obj/machinery/vending/Topic(href, href_list)
+/obj/machinery/vending/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, \
+															datum/tgui/master_ui = null, datum/ui_state/state = physical_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "vending", name, width, height, master_ui, state)
+		ui.open()
+
+/obj/machinery/vending/ui_data()

```

Comment:
This proc needs to use the lazy helpers

Please do not initialize lists directly on the object, or in `/new` please use the `LAZY_` list helpers when working with lists.

More about that can be found [here](https://github.com/HippieStation/HippieStation13/blob/master/code/__HELPERS/lists.dm#L459)

and

Please use _istypeless for loops_ you can find more information about this in the [contributing.md](https://github.com/HippieStation/HippieStation13/blob/master/CONTRIBUTING.md)